sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveBloomFilterJoin.scala

```diff
@@ -0,0 +1,109 @@
+/**
+ * Insert a bloom filter on one side of the join if it may be spill when sorting and
+ * the other side less than 100000000L rows.
+ */
+case class AdaptiveBloomFilterJoin(sparkSession: SparkSession)
+    extends Rule[LogicalPlan] with JoinSelectionHelper {
+
+  // Larger number of items may take a long time to build bloom filter.
+  private final val maxNumItems = 100000000L
+  // The factor of raw data to Java objects.
+  private final val factor = 2
+
+  private val sc = sparkSession.sparkContext
+  private val fraction = sc.conf.get(config.MEMORY_FRACTION)
+  private val storageFraction = sc.conf.get(config.MEMORY_STORAGE_FRACTION)
+  private val executorCores = sc.conf.get(config.EXECUTOR_CORES)
+
+  private val memoryPerTask =
+    (sc.executorMemory * fraction * (1 - storageFraction) * (1L << 20)).toFloat / executorCores
+
+  private def canPruneLeft(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftSemi | RightOuter => true
+    case _ => false
+  }
+
+  private def canPruneRight(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftSemi | LeftOuter => true
+    case _ => false
+  }
+
+  private def muchSmaller(filterSide: BigInt, pruningSide: BigInt): Boolean =
+    filterSide < maxNumItems && filterSide * 5 < pruningSide
+
+  private def avgSizePerPartition(logicalPlan: LogicalPlan): Float =
+    logicalPlan.stats.sizeInBytes.toFloat / conf.numShufflePartitions
+
+  private def nonBroadcastHashJoin(join: Join): Boolean = {
+    !canPlanAsBroadcastHashJoin(join, conf) && join.children.forall {
+      case LogicalQueryStage(_, stage: ShuffleQueryStageExec) => stage.isMaterialized
+      case _ => false
+    }
+  }
+
+  private def insertPredicate(
+      pruningKeys: Seq[Expression],
+      pruningPlan: LogicalPlan,
+      filteringKey: Seq[Expression],
+      filteringPlan: LogicalPlan): LogicalPlan = {
+    val filteringRowCount = filteringPlan.stats.rowCount.get
+    // To improve build bloom filter performance.
+    val coalesceNum = scala.math.ceil(filteringRowCount.toDouble / 4000000).toInt
+
+    val bloomFilterAgg =
+      new BloomFilterAggregate(new XxHash64(filteringKey), Literal(filteringRowCount.toLong))
+    val alias = Alias(bloomFilterAgg.toAggregateExpression(), "bloomFilter")()
+    val aggregate = ConstantFolding(Aggregate(Nil, Seq(alias),
+      Repartition(coalesceNum, false, filteringPlan)))
+
+    val bloomFilterSubquery = ScalarSubquery(aggregate, Nil)
+    Filter(BloomFilterMightContain(bloomFilterSubquery, new XxHash64(pruningKeys)), pruningPlan)
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDown {
+    case join @ ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, _, _, left, right, _)
+        if left.stats.isRuntime && right.stats.isRuntime && nonBroadcastHashJoin(join) =>
+      val leftRowCnt = left.stats.rowCount.get
+      val rightRowCnt = right.stats.rowCount.get
+
+      if (canPruneLeft(joinType) && muchSmaller(rightRowCnt, leftRowCnt) &&
```

Comment:
We should check the filterSide. rowCount  > 0 ?

Reply:
will [AQEPropagateEmptyRelation](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEPropagateEmptyRelation.scala) not take care of empty side ? This rule runs before Adaptive BFJ rule ?